OpenRA.Game/GlobalChat.cs

```diff
@@ -0,0 +1,371 @@
+			}) { Name = "IrcListenThread" }.Start();
+		}
+
+		void AddNotification(string text)
+		{
+			var message = new ChatMessage(DateTime.Now, ChatMessageType.Notification, null, text);
+			Game.RunAfterTick(() =>
+			{
+				History.Add(message);
+				Log.Write("irc", text);
+			});
+		}
+
+		void AddMessage(string nick, string text)
+		{
+			var message = new ChatMessage(DateTime.Now, ChatMessageType.Message, nick, text);
+			Game.RunAfterTick(() =>
+			{
+				History.Add(message);
+				Log.Write("irc", text);
+			});
+		}
+
+		void OnConnecting(object sender, EventArgs e)
+		{
+			AddNotification("Connecting to {0}:{1}...".F(Game.Settings.Chat.Hostname, Game.Settings.Chat.Port));
+			connectionStatus = ChatConnectionStatus.Connecting;
+		}
+
+		void OnConnected(object sender, EventArgs e)
+		{
+			AddNotification("Connected.");
+			connectionStatus = ChatConnectionStatus.Connected;
+
+			// Guard against settings.yaml modification
+			var nick = SanitizedName(Game.Settings.Chat.Nickname);
+			if (nick != Game.Settings.Chat.Nickname)
+				Game.RunAfterTick(() => Game.Settings.Chat.Nickname = nick);
+
+			client.Login(nick, "in-game IRC client", 0, "OpenRA");
+			client.RfcJoin("#" + Game.Settings.Chat.Channel);
+		}
+
+		void OnDisconnecting(object sender, EventArgs e)
+		{
+			if (connectionStatus != ChatConnectionStatus.Error)
+				connectionStatus = ChatConnectionStatus.Disconnecting;
+		}
+
+		void OnDisconnected(object sender, EventArgs e)
+		{
+			Game.RunAfterTick(Users.Clear);
+
+			// Keep the chat window open if there is an error
+			// It will be cleared by the Disconnect button
+			if (connectionStatus != ChatConnectionStatus.Error)
+			{
+				Game.RunAfterTick(History.Clear);
+				topic = null;
+				connectionStatus = ChatConnectionStatus.Disconnected;
+			}
+		}
+
+		void OnError(object sender, ErrorEventArgs e)
+		{
+			// Ignore any errors that happen during disconnect
+			if (connectionStatus != ChatConnectionStatus.Disconnecting)
+			{
+				connectionStatus = ChatConnectionStatus.Error;
+				AddNotification("Error: " + e.ErrorMessage);
+			}
+		}
+
+		void OnKick(object sender, KickEventArgs e)
+		{
+			Disconnect();
+			connectionStatus = ChatConnectionStatus.Error;
+			AddNotification("Error: You were kicked from the chat by {0}".F(e.Who));
+		}
+
+		void OnJoin(object sender, JoinEventArgs e)
+		{
+			if (e.Who == client.Nickname || e.Channel != channel.Name)
+				return;
+
+			AddNotification("{0} joined the chat.".F(e.Who));
+			Game.RunAfterTick(() => Users.Add(e.Who, new ChatUser(e.Who, false, false)));
+		}
+
+		void OnChannelActiveSynced(object sender, IrcEventArgs e)
+		{
+			channel = client.GetChannel(e.Data.Channel);
+			AddNotification("{0} users online".F(channel.Users.Count));
+			connectionStatus = ChatConnectionStatus.Joined;
+
+			foreach (DictionaryEntry user in channel.Users)
+			{
+				var u = (ChannelUser)user.Value;
+				Game.RunAfterTick(() => Users.Add(u.Nick, new ChatUser(u.Nick, u.IsOp, u.IsVoice)));
```

Comment:
Looks like it only happens that once. At least it doesn't happen when users join or leave. Could be that it's run once every hour or so, but I didn't test that long.